### PR TITLE
Move error handling from  to  transform

### PR DESCRIPTION
**Pull request by FreeMasen**
This PR move the error handling from the exiting `output` transform function into a new `error` transform. 

~The primary driver here is that MacOS will return `nil, "timeout"` on calls to `connect` but when write readiness is reached to complete the connection calling `connect` again will return `nil, "already connected"`. For good measure I've added this same error handling to `accept` though it may be removable.~

After additional investigation the `nil, "already connected"` error on macos is a fatal error. A different mechanism is going to need to be introduced to handle functions that shouldn't be considered "successful" after the call to `coroutine.yield` resolves.

I believe this closes #36

### CODE REVIEW

**Comment by FreeMasen** [comment]
It seems that @azdle is no longer hosting `bin.azdle.net` so I've replaced that with a locally running server just for the `http*` tests. The server can be found [here](https://github.com/cosock/test-http-server), which I believe is doing exactly the same thing as previously. This does complicate the `scripts/run_all_tests.sh` I've added here but it should be possible to get that working again w/o too much hassle if folks feel like it is important

File: cosock/socket/tcp.lua [hunk 1]
```diff
@@ -92,22 +92,22 @@ m.receive = passthrough("receive", function()
...
-
-      if err then
-        return nil, err, data
-      else
-        return data
+      return nil, err, table.concat(parts)
+    end,
+    -- transform output after final success
+    output = function(recv)
+      assert(not (recv and partial), "socket receive returned both data and partial data")
```
Comment (NoahCornell) [inline cosock/socket/tcp.lua:123]: `partial` is no longer provided in this function

File: cosock/ssl.lua [hunk 2]
```diff
@@ -48,21 +48,24 @@ m.loadcertificate = passthrough("loadcertificate")
 m.newcontext = passthrough("newcontext")
 
 m.receive = passthrough("receive", {
-  output = function(bytes, err, ...)
+  output = function(bytes)
+    return bytes
+  end,
+  error = function(_sock, err, ...)
```
Comment (NoahCornell) [inline cosock/ssl.lua:54]: Do we need to be handling partial receives here too?

**Review by NoahCornell — APPROVED** [review]

File: cosock/ssl.lua (hunk 2 above)
Comment (FreeMasen) [reply cosock/ssl.lua:54]: This is an interesting question, I am a little surprised that the same scheme isn't required here as is defined in the `tcp.lua` code. I suspect that the wrapping operation means that the `tcp.lua` transform will be applied before we reach this level meaning that the `partial` will be included as the return on line 58 (`...`) if one was provided

**Review by cjswedes — APPROVED** [review]

File: cosock/ssl.lua (hunk 2 above)
Comment (FreeMasen) [reply cosock/ssl.lua]: holy cow, this has apparently been a bug since forever, if an ssl socket yields with a partial payload it will end up dropping that data on the floor. I've added a new test to validate this (mostly a copy of the tcp large-payload test) and refactored the tcp receive transform to be part of the module export so that the ssl module can use the same implementation

Comment (NoahCornell) [reply cosock/ssl.lua]: Yeah I think we fixed it in scripting-engine but I wasn't sure if the behavior was different here for some reason

File: cosock/socket/tcp.lua [hunk 3]
```diff
@@ -40,7 +40,14 @@ end
...
 
-m.connect = passthrough("connect")
+m.connect = passthrough("connect", {
+  error = function(self, err)
+    if err == "already connected" then
+      return self
+    end
+    return nil, err
+  end
+})
```
Comment (azdle) [inline cosock/socket/tcp.lua:122]: I'm not sure about adding this check. Are you seeing cases where this error is being spuriously returned?

Also, if we do want it, I think it should return `1.0` not `self`, unless I'm missing something.

File: cosock/socket/tcp.lua [hunk 4]
```diff
@@ -24,13 +24,24 @@ end})
...
+      return ctor(inner_sock)
+    end,
+    error = function(inner_sock, err, ...)
+      if err == "already connected" then
+        return ctor(inner_sock)
+      end
+      return nil, err, ...
+    end,
+  }
+end)
```
Comment (azdle) [inline cosock/socket/tcp.lua:105]: Did you mean to do this on `connect` and not `accept`?

File: scripts/run_all_tests.sh [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+REPO_ROOT="$(dirname $(dirname "$(realpath "${BASH_SOURCE:-$0}")"))"
+
+# install the test server binary in the ./bin directory
+if ! command -v ./bin/cosock-test-server 2>&1 >/dev/null
+then
+    export COSOCK_TEST_SERVER_INSTALL_DIR="${REPO_ROOT}"
+    curl --proto '=https' --tlsv1.2 -LsSf \
+      https://github.com/cosock/test-http-server/releases/download/v0.1.7/cosock-test-server-installer.sh \
+      | sh
+fi
```
Comment (azdle) [inline scripts/run_all_tests.sh:13]: Not sure how I feel about automatically installing something especially via piping curl to sh.

I've got `bin.azdle.net` back up for now. Could we remove the server changes from this PR? And then either put them in a separate PR or I could hack together a little lua script that does the whole delay thing so that there's no other deps needed.

File: cosock/socket/tcp.lua (hunk 3 above)
Comment (FreeMasen) [reply cosock/socket/tcp.lua:122]: I think I am going to remove the macos fixes from this PR and add them to their own
